### DISCUSSION TITLE
Please convince me to keep using Genmon. Tell me something is wrong.

### DISCUSSION THREAD

**Discussion by jasonrusso7** [jgyates/genmon#1104] [General]
I got a preloaded Raspberry Pi in December.  To say it has been frustrating is a gross understatement.  I have spent DAYS of my life messing with this thing.  Instead of giving me piece of mind that my generator is monitored, it has been one more thing that I need to CONSTANTLY babysit.   At first, it would stay connected for 1 day at the most.  Every day the green LED would be solid indicating that it was overloaded. It was so bad that I wired in a toggle switch to disconnect the power to reboot it several times a week.  This went on for a few months.  

Working with @pintsizeme we figured out the SD card was bad.  I wrote a new SD card and figured out how to install Genmon (defeating the purpose of buying a preloaded unit).  The made it better, but still not good.  Now it was lasting maybe 3-4 days.  This week I started getting another error and I'm back to losing connection every few hours.  

I want to monitor Genmon through my ISY (because I have no faith the program is running and will send me notifications), so I installed nodelink on a mini PC. I got everything working but it is so unreliable I am constantly checking.  Between the Pi losing connection and nodelink needing to be restarted it has been an awful experience.  Yesterday I installed Home Assistant in a virtual Linux PC with the hopes of monitoring it that way, but that is now turning into a headache.  I can't access the HA console and the Pi is frozen up again. 

I have the Generac Mobile link on my table, but I hate to admit defeat. I also don't want to spend another $180 and $50 a year.  

Please tell me it gets better and something is wrong, because I can't imagine that this community tolerates this level of aggravation.

**Comment by djstroh** [comment]
I don’t use a PI. I used HA, esphome with a serial bridge (Esp32), and run
genmon on the same machine as HA. Very stable and I have all the cpu power
needed.

Hope that helps.

**Reply by jasonrusso7** [reply]
Here is my problem now, I have Home Assistant running in a virtual box, but my browser doesn't recognize the IP address so I can't access it.

How did you make the connection with the serial bridge? Do you have instructions?

How are you running Genmon? Is it on a PC or do you have a stand alone PC with HA (not a virtural) and Genmon.

**Reply by jgyates** [reply]
The instructions for using a serial bridge with genmon is in the wiki: https://github.com/jgyates/genmon/wiki/Appendix-F----Serial-over-IP

The above is for the genmon setup for serial over TCP (serial bridge), the page contains a link for ESP based serial bridges or you can buy one from a retailer (links included in the wiki page).

Feel free to use any platform you like however most people do use the pi with success. Based on the feedback and logs submitted  over 98 percent of genmon use is with the pi. Having said that there are folks who use virtual environments and stand alone PCs with success.

If you wifi signal is marginal you will likely have another set of issues if you opt to use serial over wiif. If marginal wifi is your issue there are a few things you can do to make this issues not as bad if you are using serial over wiif, but the best solution is to either use a Ethernet connection to the pi, or a longer serial connection and power the pi closer to you house. RS-232 has a limit of around 50 feet, but some folks use RS-422 or RS-485 for longer cable lengths.

**Reply by jasonrusso7** [reply]
How can I use an ethernet connection if there is no ethernet port? I would love to hardwire.  Is there an adapter? 

Still unsure how he is running Genmon and Home Assist on the same machine.

**Reply by jgyates** [reply]
Most raspberry pis have an Ethernet port. I believe the ones that pintsize.me ship have an Ethernet port as well. In this scenario you would connect the pi to the controller via serial and connect the pi to your network via Ethernet. Alternatively you can use a serial over TCP bridge. In that situation you would connect serial from the bridge to the controller and, depending on the bridge you purchase the bridge would support Ethernet or wifi for the network so you could connect the bridge to your network via Ethernet. Then you can run genmon on a pi or some other linux based environment and it will communicate to the serial bridge via TCP over Ethernet.

If you are not sure you can send a picture of your pi and that will confirm if you have Ethernet on the pi. Of course you may need to drill a hole in the enclosure to get to the Ethernet port if one has not been drilled prior.

**Reply by skipfire** [reply]
I haven't had any B's for awhile, just 3A+ and Zero 2 (now I'm down to just Zero 2s).

**Reply by jasonrusso7** [reply]
I have a 3A+

**Reply by robybrisson** [reply]
> I don’t use a PI. I used HA, esphome with a serial bridge (Esp32), and run genmon on the same machine as HA. Very stable and I have all the cpu power needed. Hope that helps.
> […](#)

@djstroh 
Could you elaborate, for a newb, how you've set this up?  I have followed https://github.com/gregmac/Genmon-ESP32-Serial-Bridge?tab=readme-ov-file ESP32 hardware configuration.  I was actually hoping I could do a direct ESPHome to HA setup with https://esphome.io/components/modbus_controller.html to avoid running another device, but I'm not sure I have the technical chops to decipher jgyates code into yaml configs, or if that would even work with the ESP32 HW setup from gregmac.

I've got HAOS running on a x86 device with I5/16gb of ram so I have plenty of power, but I don't know how to get genmon running side-by-side with HA.

Thanks for any assistance.

**Comment by jgyates** [comment]
@jasonrusso7 

Something is wrong. This is not normal to have to babysit the software or a raspberry pi unless something is wrong somewhere. 

First, let me get a good idea of what is working and what is not working. When you say the pi is disconnected, are you referring to the wifi/network, serial or is the pi not staying powered up and accessible? 

You also reference a green LED, is this the LED on the controller or on an enclosure for the pi? If it is an LED for the enclosure I am assuming this is one that came with pintsize.me enclosure? I have used the HAT, but I don't have an enclosure from them. @skipfire is the pintsize.me guy. 

A few questions:
1) What is your generator model?
2) how are you powering the pi? Are you using the cables from pintsize.me to power the pi from from the controller, or are you powering the pi via the battery?
3) How are you communicating with the controller, direct serial or serial over TCP?
4) what is the network connection to the pi, wifi or Ethernet? 
5) do you have outboud email setup and working with genmon? if your email is configured and working then you can submit your log files on the About page and I can look at them to see if anything is obviously wrong.

Assuming you connection issues are network related and you are using wifi can you look on the Monitor page and get the WLAN signal level, signal quality and signal noise? A marginal wifi signal can cause intermittent network issues.

**Reply by jasonrusso7** [reply]
I should be more clear.  I say disconnected when I can't access the Pi.  At first it was always a solid green led ON THE PI indicating 100% activity.  Now, it just cannot be accessed either through SSH or VNC.  Cycling the power brings it back online, so I wouldn't think it's a network issue because that didn't change.

I've tried multiple networks, 2.4ghz and 5ghz. It always seems like it's a weak signal even though I have an outdoor wireless access point 20 feet away.  I can connect to the access point with my laptop inside the house and it seems like the internet is fine.  

1.   2012 Generac with a Nexus controller
2.   Using the wiring to the 12V battery and the barrel connector. 
3.   Not sure what you mean, I am using the molex connector that came with the setup
4.   Wifi, there is no ethernet plug but I would be happy to hardwire it if possible. I plan to hardwire the generac unit if I go that way.
5.   I have never been able to get the email to work, but I can download the logs when I get home and power cycle.....again. 

If it is in fact a signal issue, and 20 feet from an access point is too far, is there a way to hardwire it? Like I said, there is no ethernet port.

**Reply by jgyates** [reply]
If you put the pi enclosure inside the generator enclosure then the metal generator encolsure blocks a lot of the wifi signal. What signal does make it through has has a lot of reflection / noise in the signal as it is being bounced off a lot of things to get to the pi through the enclosure. Some people add an external wifi antenna routed through the metal enclosure, but Ethernet is a better solution in my view.

also, I think the older models (Nexus) and liquid cooled models have thicker metal enclosure walls so wifi has a harder time penetrating. I have an Evolution Liquid cooled and I had issues with wifi signal quality, even with outdoor antennas. I ended up with a direct Ethernet connection. Just make sure it is either in a conduit or you use outdoor Ethernet cable. Exposed non-outdoor Ethernet cable will deteriorate over time if exposed to the elements, especially heat.

**Reply by jasonrusso7** [reply]
I HAD it inside the generator and the signal was very weak. I moved it outside and it was better.  I confirmed that the Pi does not have ethernet, but would a USB-ethernet adapter work?   Would it need drivers to load? 

I have outdoor - direct burial cat 5E.

**Reply by jgyates** [reply]
Ethernet should work.

https://www.amazon.com/Smays-Micro-B-Ethernet-compatible-Raspberry/dp/B01AT4C3KQ/
 or
https://shop.pimoroni.com/products/three-port-usb-hub-with-ethernet-and-microb-connector?variant=22477883847

should work with the pi zero 2, which sounds like what you have.

**Reply by jgyates** [reply]
Just check the USB connector before you purchase. I am pretty sure it uses a micor USB like the USB-Ethernet adpaters above do, but I don't have a pi zero 2 handy to validate.

**Reply by jasonrusso7** [reply]
I have a 3A+ which i just looked up has a micro-usb, but it says for power.  If I am using the barrel connector, will the USB port support data?

**Reply by jgyates** [reply]
The 3A+ has one micro USB for power and one USB-A for peripheral support. I would use the USB-A for USB Ethernet and get an adapter that interfaces USB-A. The micro USB on the 3A+ does not support data.

**Reply by jasonrusso7** [reply]
I attached the log file.  I also attached the screenshot of the genmon console.  The wifi is weak even though its 15-20 feet away. Hopefully the usb dongle works as advertised.  I'm not sure why it gets so hot running at such a low cpu load. 

Tonight it didn't even stay online for more than a couple hours.  If it were a weak network signal, shouldn't the Pi reconnect itself? I power cycled the Pi again and it came right back.  Is it bad hardware??  I've spent enough frustrating weekends. If it is just bad hardware let me know. 

[genmon_logs(1).tar.gz](https://github.com/jgyates/genmon/files/15018354/genmon_logs.1.tar.gz)
![image](https://github.com/jgyates/genmon/assets/167337356/ce9819ef-b87a-447c-9ece-156b508a008a)
![image](https://github.com/jgyates/genmon/assets/167337356/e45de2b0-c611-40e7-a741-ed3fe1cd30b2)

**Reply by jasonrusso7** [reply]
aaaaaand it just cut out again after 5 minutes.  It doesn't show up in the access point anymore. 

I'm running a hard line out to the generator on Saturday. It's either getting plugged into the Pi or the Generac unit.  I'd rather the Pi, but I can't take more of this.

**Reply by jgyates** [reply]
your communication stats and platform stats look normal and indicate that genmon is working. You screen capture of the Monitor page does not show the wifi signal info. You logs look good from a serial perspective but there are indications that your network is the problem so this backs up the bad wifi signal theory.

**Reply by jasonrusso7** [reply]
I showed the screen shot of the wifi gauge, it's in the red. Could it be a bad wifi transmitter on the Pi itself?  I have wifi hundreds of feet from my house, the Pi is 20 feet away. If that's the case, the hardwire would take care of it. Wouldn't the Pi reconnect to wifi?

**Reply by jgyates** [reply]
I guess it is possible the pi could be defective. If you had another pi you could try it out, but I agree, the wired will solve this issue. The pi OS is not great about reconnecting. A lot of people use a script to detect if the connection is dropped and then the script will attempt to reconnect. The tips and tricks wiki page has info on the script, that may help the connection, but my guess is that you will still have issues:

https://github.com/jgyates/genmon/wiki/Appendix-M-Tips-and-Tricks-useful-for-Genmon#setting-up-a-script-to-check-the-wifi-connection-and-reconnect-if-needed

The script is recommended to be run ever 2 - 5 minutes to check if your wifi is down and then it will reconnect so you would still have periodic connection issues, they just may not last as long (shorter dropouts, but probably more frequent occurrences if you are losing connection due to a weak signal).

**Comment by djstroh** [comment]
I would also monitor to see how hot the CPU is getting. Having intermittent
hardware failures usually tells me things are getting too warm.  Make sure
you keep it cool below the 60 degree C value.

**Reply by jasonrusso7** [reply]
The other day it was getting up to about 68. That's when I started getting the warning. The ambient temp was about 18 degrees so that may be a problem in the future.

**Reply by jasonrusso7** [reply]
Can you elaborate on my questions on the prior post?  I'm interested in your setup

**Comment by djstroh** [comment]
I followed this setup "https://github.com/gregmac/Genmon-ESP32-Serial-Bridge"
and used an external antenna on the esp32 to get a better wifi signal. I
use docker on a Ubuntu system to run genmon, HA and esphome.

A few things that could affect your wifi strength. 1) What type of siding
do you have? 2) What type of wifi equipment? I am using a tp-link ax3000 AP
that has a directional antenna to talk to the ESP32.

With the symptoms you are having I would try some different hardware. You
just might have a bad board. Make sure everything stays dry and cool. Note
time of day things are failing, is it the hottest part of the day.

**Reply by jasonrusso7** [reply]
Unfortunately this is my first venture ever with Raspberry Pi so I have nothing to compare it to.  Just problems after problems thus far.

**Reply by jasonrusso7** [reply]
@djstroh Is there a tutorial on installing genmon on Home Assistant?

**Comment by jasonrusso7** [comment]
I got this one.    Reviews say it's plug and play on a 3A+
  

  
Cable Matters Plug  &  Play USB to Ethernet Adapter with PXE, MAC Address Clone Support (Ethernet to USB 2.0 Adapter, Ethernet Adapter for Laptop) Supporting 10/100 Mbps Ethernet Network Black https://a.co/d/cWcIton
  

  

  
  
Thanks,
  

  
Jason
  

  

        
  
  
>

**Comment by djstroh** [comment]
It is in the link that I sent you on GitHub. You have to copy some files
over in HA.

**Reply by jasonrusso7** [reply]
This stuff makes me feel dumb. Are there tutorials on how to write code in this?  I am pretty sure I put the generator.yaml file in the right place, but the dashboard data says there is something wrong.  

Do I need to make any changes in the genmon dashboard?

**Comment by jasonrusso7** [comment]
Cool, I'll look through it   
  

  

  
  
Thanks,
  

  
Jason
  

  

        
  
  
>

**Comment by Jkh62** [comment]
I am running a Pi Zero W and my Wi-Fi was weak  ( yellow range ) but not in the red like yours in the pic.
 My router is set to  20 MHz only / 2.4Ghz band and I went thru the channels to find the strongest one and it moved it into the solid green area and now never drops into the yellow range.  It was a single cup of coffee job.

Set a static IP address for your pi.

My install is still fairly new but has been running 24X7 without any issues for about 50 days straight.

Its worth the effort to get it up and running correctly and some very sharp people here to assist.

**Reply by jasonrusso7** [reply]
50 days straight? No reboots? That's good to know it's possible.

I don't want to start messing with channels. I have 80 devices on my Wifi 6 mesh network.

My plan is to run a ethernet to the Pi after installing a dongle to get the wifi out of the equation. If that doesn't work, then I will swap it out for a 3B (I have a 3A+) that has on board ethernet.

I'm not giving up yet. I figured out how to run virtural Home Assistant, now I'm trying to figure out how to integrate genmon into it.

**Reply by spatel7786** [reply]
Im going on 100 days, 24/7 non stop using a RPI3B+ with a pintsizeme HAT and pintsizeme Splitter to also keep the generac wifi app working  for backup monitoring. NO issues what-so-ever! it just .....WORKS. Im also using wifi on the RPI and its not hardwired. 

![image](https://github.com/jgyates/genmon/assets/158606188/1d3bf400-5a84-4e98-8b81-110f49910586)

**Reply by jasonrusso7** [reply]
100 days, that's great. I think that we have the same exact Pi. Where and how do you have it mounted?  I have it in a waterproof enclosure and the temperature started throttling the Pi the other day. Your temp is a bit high as well.

**Reply by spatel7786** [reply]
Mine is also in a water proof enclosure sitting on top of the battery inside the generac. I noticed the temps getting to 180 degrees one day so I decided to remove the foam interior from the box and leave the enclosure cracked open a bit. the open side is facing inwards away from the louvers, keeping things out of the enclosure. This is keeping the Pi cooler than before.

**Comment by Jkh62** [comment]
The only set up issues I had were self imposed from security settings on my network and flipping the correct dip switch on the HAT because I did not read the instructions. Hard wired should be the best connection but my WIFI has been 100% thus far.

**Comment by djstroh** [comment]
I don’t remember changing the yaml file. You do need to setup Mqtt and an
Mqtt server. Enable the mosquito one in HA OS.

I have not found any tutorial on setting that up. Just a few comments on
the GitHub page.

**Comment by jgyates** [comment]
@jasonrusso7, were you able to get genmon working reliably with a wired connection?

**Reply by jasonrusso7** [reply]
It seems to have!!  The temperature is still a little weird, but it has been much much more stable.  I had to reboot it after a week because I was getting no data from generator warnings.  After the reboot, I haven't had any issue. 

I'm still trying to figure out how to integrate it with Home Assistant.  I'm pretty sure I put the yaml data in the right place, but I can't figure out the dashboard. 

Thanks for the help!!!!

**Comment by djstroh** [comment]
Try creating a new dashboard, enter edit dashboard (upper right), enter raw
configuration editor (upper right). I then cut and pasted the
dashboard into the raw configuration editor.

**Reply by jasonrusso7** [reply]
This is what I did. 

![image](https://github.com/jgyates/genmon/assets/167337356/8c96c1b5-27db-473c-8e0e-0d32b1138326)

![image](https://github.com/jgyates/genmon/assets/167337356/4b7f4d73-4b55-4f38-93d6-5d84394b322d)

**Reply by jasonrusso7** [reply]
It's probably something minor like too many indents or something.  I'm not a programmer so this isn't very intuitive for me, but I am learning.

**Reply by skipfire** [reply]
I'm not familiar with this or with yaml, but I have a few guesses you could try.
1. Dedent icon line 2 spaces or remove entirely
2. Maybe remove the line for `title: Generator` in case it's a duplicate issue
3. Maybe you have some mixing of spaces and tabs for indenting

**Comment by djstroh** [comment]
The first few lines of my yaml look like:
views:
  - cards:
      - type: vertical-stack
        cards:
          - type: custom:mushroom-title-card
            title: Generator Status
          - type: custom:mushroom-template-card
            primary: '{{ states(''sensor.generator_engine_state'') }}'
            secondary: ''


Yaml code needs a very precise indent and make sure you have no tabs!

**Comment by djstroh** [comment]
I am going to share with you my setup. I built my own ESP32 device with an
RS232 device on it. I will attach my esphome yaml file.  It uses a stream
server to communicate with the docker genmon module. It has been running
for over a year without any issues.

I run gemon under a docker container. I will attach the Dockerfile I use to
load it. Below is the docker-compose entry.

  genmon:
    build: genmon
    container_name: genmon
    restart: unless-stopped
    environment:
      - TZ=America/Los_Angeles
    ports:
      - "8000:8000" # port for web interface
    volumes:
      - /opt/genmon:/etc/genmon
